### PR TITLE
Replace local copyright check with pre-commit-hooks verify-copyright

### PR DESCRIPTION
**Pull request by KyleFromNVIDIA**
## Description
The local `copyright.py` script is bug-prone. Replace it with a more robust centralized script from `pre-commit-hooks`.

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by karthikeyann** [comment]
Are the verify-copyright failures due to the changes to files currently in this PR?

**Comment by KyleFromNVIDIA** [comment]
Yes. It appears there's more work to do on the new `verify-copyright` check. Closing for now.

**Comment by KyleFromNVIDIA** [comment]
> Eg. join.pxd is updated to 2021, but it expect it to be 2024.

This new script is more intelligent about figuring out whether or not a file needs to be updated, and what year it should be updated to. The `--batch` mode will retroactively update files to the last time they were changed. Prior to this PR, `join.pxd` was last updated in 2021, so `--batch` mode updates it to 2021.

**Comment by PointKernel** [comment]
@KyleFromNVIDIA Nice work!

Will the new script work for a non-rapids repo? I'm planning to apply it to https://github.com/NVIDIA/cuCollections as well.

**Comment by KyleFromNVIDIA** [comment]
The script is currently targeted specifically at RAPIDS repos. We have [plans](https://github.com/rapidsai/pre-commit-hooks/pull/9#discussion_r1462392086) to make it more configurable for non-RAPIDS repos in the future.

In any case, I have some more work to do on this, so don't merge it yet.

**Comment by bdice** [comment]
@KyleFromNVIDIA I'll mark this as a draft for now. Happy to review again when it's ready.

**Comment by gerashegalov** [comment]
> The script is currently targeted specifically at RAPIDS repos. We have [plans](https://github.com/rapidsai/pre-commit-hooks/pull/9#discussion_r1462392086) to make it more configurable for non-RAPIDS repos in the future.

Consider adding support for NVIDIA/spark-rapids-* repos as well: NVIDIA/spark-rapids-benchmarks#181

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -154,6 +148,10 @@ repos:
     hooks:
       - id: ruff
         files: python/.*$
+  - repo: https://github.com/rapidsai/pre-commit-hooks
+    rev: 253ff58a8a842a517995a7cdf6832aff20356b2a
```
Comment (bdice) [inline .pre-commit-config.yaml:152]: I'd like to adopt some kind of version/tag strategy for this repo, so we don't have to pin to hashes. https://github.com/rapidsai/pre-commit-hooks/issues/16

**Review by bdice — APPROVED** [review]

File: .pre-commit-config.yaml (hunk 1 above)
Comment (bdice) [reply .pre-commit-config.yaml:152]: I'd tag the current state of that repo as `v0.0.1` and move forward with that tag. Either way is fine but I suspect there will be a lot of other repos who will copy-paste this from cudf and we probably don't want them to use a hash.

Comment (KyleFromNVIDIA) [reply .pre-commit-config.yaml]: Created `v0.0.1` and updated to use it.

**Comment by bdice** [comment]
@KyleFromNVIDIA Is anything more needed here? Please ping ops for a review in Slack if that's the only missing piece.

**Review by AyodeAwe — APPROVED** [review]

**Comment by karthikeyann** [comment]
Will the new pre-commit hook update correctly for PRs created in past year, but still open this year? I had a similar issue with old copyright script in local pre-commit runs. Old script updated a few files which are not part of the PR. 
If it is not possible to test or deduce it, it may be ignored.

**Review by karthikeyann — APPROVED** [review]
Great work @KyleFromNVIDIA  👍

**Review by wence- — APPROVED** [review]

**Comment by harrism** [comment]
Any idea why this is failing?

**Comment by bdice** [comment]
> I'm running CI again.

@KyleFromNVIDIA This failed again because clicking the "Rerun failed jobs" button does not update the `shared-workflows`. It uses the `shared-workflows` as of the time of that commit. To get the changes from https://github.com/rapidsai/shared-workflows/pull/182 that are needed to make CI pass, you have to make a new commit (merge the upstream or push an empty commit). I'll merge the upstream.

**Pull request merged**